### PR TITLE
feat(runtime): make it possible to yield and resume within the same block

### PR DESCRIPTION
**Pull request by jancionear**
Fixes case 2) of https://github.com/near/nearcore/issues/14904

### The problem

There is an issue in yield/resume, where executing `yield` and `resume` within the same block doesn't work. This happens because `yield` generates an outgoing `PromiseYield` receipt, which is applied at the next height as an incoming receipt. Trying to resume before the `PromiseYield` receipt is applied results in failure, so resuming in the same block as the `yield` doesn't work, it needs to be at least one block later.

This causes issues for users who expect that calling `resume` after `yield` will always work, and then get a nasty surprise when suddenly it doesn't work if the transactions which run `yield` and `resume` end up being executed in the same block. See https://near.zulipchat.com/#narrow/channel/295558-core/topic/Yield.2FResume.20edge.20case/with/560809761 for the discussion about this issue.

### The Fix

To fix the problem, let's apply the `PromiseYield` receipts immediately after they're produced, without sending them out as outgoing receipts.
To achieve this, in https://github.com/near/nearcore/pull/14937 I added a concept of "instant receipts" to the runtime. Instant receipts are receipts which should be applied immediately after they are created. Then we can make `PromiseYield` receipts into instant receipts to apply them immediately.
This felt cleaner and less error prone than trying to call `set_promise_yield_receipt`  somewhere in the middle of `apply_action_receipt`. I'm also hoping that the concept of instant receipts will be useful for the "Instant delete" feature that was discussed recently.

The fix shouldn't break any contracts because there was never a guarantee whether two transactions/receipts would execute in one block or not. Even if the user tries to submit two transactions at the same time, one of them could be postponed to delayed receipts. Valid contracts had to handle both the case where two receipts are executed in the same block and the case where they're executed in two different blocks, handling how yield/resume behaves in both cases. Now we will make only one of those behaviors possible.

The only way I can imagine it being a breaking change is if someone checks the block height where a `FunctionCall` is executed, compares across two receipts and asserts yield/resume behavior based on that. Then they'd be able to detect that the behavior has changed. That seems unlikely, but I'll ask around about it just to be sure.

### Tests

The existing tests should hopefully ensure that yield/resume works fine after the change.
I added a test which reproduces this corner case. As expected, the resume fails before the fix and works fine after.

It would be nice to rewrite all of the yield/resume tests from `TestEnv` to `TestLoop`, but that'd require some work. For now I built on top of the existing tests.

---

The PR is divided into commits for easier review.

### CODE REVIEW

**Comment by jancionear** [comment]
> Just FYI instant receipts won't be handled by the indexer, so those won't be visible in explorers. I will fix that in a separate PR.

Ah good point, I made sure that the execution outcome is saved by `process_receipt_with_metrics`, but I forgot about the receipt itself.
It would probably make sense to change `outcomes` in `ApplyProcessingReceiptState` to be `Vec<ExecutionOutcomeWithReceipt>`. Then we would get a list of all processed receipts and their outcomes as the output of chunk application and we could save that to the db. I thought about taking a stab at it, but I see that there's some fun indexer-specific logic with `DELAYED_LOCAL_RECEIPTS_CACHE` etc, that I'm not sure how to handle. I will leave that to you then, you probably have some plans about it.
It would be good to save all receipts to DB so that they are available to the jsonrpc `EXPERIMENTAL_receipt` endpoint. Having them as chunk output would also make it easier to fetch them in optimistic-block queries.

**Comment by jancionear** [comment]
> Please split it into 2 different PRs introducing instant receipts feature and enabling it for PromiseYield receipts. While having separate commits helps with the review, but the whole PR is still a single unit for the review. This way I can approve the instant receipts part without implicitly approving its integration with PromiseYield feature.

Alright, opened https://github.com/near/nearcore/pull/14937. I felt like the concept was easier to understand in the context of this PR. It also has a test that indirectly tests the instant receipts code, so I put them together. But a separate PR also works.

**Comment by jancionear** [comment]
Rebased on top of the latest master that includes https://github.com/near/nearcore/pull/14937 to keep the commit history clean.

**Comment by jancionear** [comment]
> Avoid adding new tests to integration-tests and using test env, we want to deprecate that. Add single node test-loop test instead, that should be pretty straightforward.

I'll look into porting the test to testloop, but for now I've got BEE things to do.
Ideally we should probably go and port all of the yield/resume tests to testloop.

**Comment by pugachAG** [comment]
@jancionear feel free to merge with `integration-tests` for now and do test-loop migration in a separate PR or just create a better eng issue for that

File: integration-tests/src/tests/features/yield_resume.rs [hunk 1]
```diff
@@ -166,3 +182,89 @@ fn yield_then_resume() {
...
+    );
+    let resume_tx_hash = resume_transaction.get_hash();
+    assert_eq!(
+        env.rpc_handlers[0].process_tx(resume_transaction, false, false),
+        ProcessTxResponse::ValidTx
+    );
+
+    // Allow the yield create and resume to be included and processed.
+    for _ in 0..3 {
+        env.produce_block(0, next_block_height);
```
Comment (saketh-are) [inline integration-tests/src/tests/features/yield_resume.rs:262]: Might be nice to verify explicitly that the two txs end up in the same block.

File: integration-tests/src/tests/features/yield_timeouts.rs [hunk 2]
```diff
@@ -54,6 +60,50 @@ fn find_yield_data_ids_from_latest_block(env: &TestEnv) -> Vec<CryptoHash> {
     result
 }
 
+/// Read all the `PromiseYield` receipts stored in the latest state and collect their data_ids.
+pub fn get_yield_data_ids_in_latest_state(env: &TestEnv) -> Vec<CryptoHash> {
```
Comment (saketh-are) [inline integration-tests/src/tests/features/yield_timeouts.rs:64]: Might help with readability to cut this into two halves (first one gets the state root, second one iterates the PromiseYield receipta for given state root).

File: integration-tests/src/tests/features/yield_resume.rs (hunk 1 above)
Comment (jancionear) [reply integration-tests/src/tests/features/yield_resume.rs:262]: Added a check

File: integration-tests/src/tests/features/yield_timeouts.rs (hunk 2 above)
Comment (jancionear) [reply integration-tests/src/tests/features/yield_timeouts.rs:64]: done

**Comment by jancionear** [comment]
@pugachAG could you clear your "requested changes"? It blocks merging.
